ingest/src/server_iot.rs

```diff
@@ -7,58 +7,268 @@ use file_store::{
     traits::MsgVerify,
     FileType,
 };
-use futures_util::TryFutureExt;
+use futures::{Stream, StreamExt, TryFutureExt};
 use helium_crypto::{Network, PublicKey};
 use helium_proto::services::poc_lora::{
-    self, LoraBeaconIngestReportV1, LoraBeaconReportReqV1, LoraBeaconReportRespV1,
-    LoraWitnessIngestReportV1, LoraWitnessReportReqV1, LoraWitnessReportRespV1,
+    self, lora_stream_request_v1::Request as StreamRequest,
+    lora_stream_response_v1::Response as StreamResponse, LoraBeaconIngestReportV1,
+    LoraBeaconReportReqV1, LoraBeaconReportRespV1, LoraStreamRequestV1, LoraStreamResponseV1,
+    LoraStreamSessionInitV1, LoraStreamSessionOfferV1, LoraWitnessIngestReportV1,
+    LoraWitnessReportReqV1, LoraWitnessReportRespV1,
 };
 use std::{convert::TryFrom, path::Path};
-use tonic::{transport, Request, Response, Status};
+use tokio::{sync::mpsc::Sender, time::Instant};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{transport, Request, Response, Status, Streaming};
 
 pub type GrpcResult<T> = std::result::Result<Response<T>, Status>;
+pub type GrpcStreamResult<T> = ReceiverStream<Result<T, Status>>;
 pub type VerifyResult<T> = std::result::Result<T, Status>;
 
+type Nonce = [u8; 32];
+
+#[derive(Debug)]
+struct StreamState {
+    beacon_report_sink: FileSinkClient,
+    witness_report_sink: FileSinkClient,
+    required_network: Network,
+    shutdown_listener: triggered::Listener,
+    pub_key_bytes: Option<Vec<u8>>,
+    session_key: Option<PublicKey>,
+    timeout: Instant,
+    nonce: Option<Nonce>,
+    session_key_offer_timeout: std::time::Duration,
+    session_key_timeout: std::time::Duration,
+}
+
+impl StreamState {
+    fn new(server: &GrpcServer) -> StreamState {
+        StreamState {
+            beacon_report_sink: server.beacon_report_sink.clone(),
```

Comment:
maybe just clone the GrpcServer struct itself into StreamState, given you are coping all the fields anyway

Reply:
my preference would be to leave it as is, as I think it is clearer what data the StreamState needs, but I am not super attached to it, so if you or others feel strongly about it, I will make the change.

Reply:
Its more of a nit, so no big deal either way